contrib/mesos/pkg/node/node.go

```diff
@@ -0,0 +1,158 @@
+		},
+	}
+	for k, v := range labels {
+		n.Labels[k] = v
+	}
+
+	// try to create
+	return client.Nodes().Create(&n)
+}
+
+// Update updates an existing node api object with new labels
+func Update(client *client.Client, n *api.Node, labels map[string]string) (*api.Node, error) {
+	patch := struct {
+		Metadata struct {
+			Labels map[string]string `json:"labels"`
+		} `json:"metadata"`
+	}{}
+	patch.Metadata.Labels = map[string]string{}
+	for k, v := range n.Labels {
+		if !IsSlaveAttributeLabel(k) {
+			patch.Metadata.Labels[k] = v
+		}
+	}
+	for k, v := range labels {
+		patch.Metadata.Labels[k] = v
+	}
+	patchJson, _ := json.Marshal(patch)
+	log.V(4).Infof("Patching labels of node %q: %v", n.Name, string(patchJson))
+	err := client.Patch(api.MergePatchType).RequestURI(n.SelfLink).Body(patchJson).Do().Error()
+	if err != nil {
+		return nil, fmt.Errorf("error updating labels of node %q: %v", n.Name, err)
+	}
+
+	newNode, err := api.Scheme.DeepCopy(n)
+	if err != nil {
+		return nil, err
+	}
+	newNode.(*api.Node).Labels = patch.Metadata.Labels
+
+	return newNode.(*api.Node), nil
+}
+
+// CreateOrUpdate tries to create a node api object or updates an already existing one
+func CreateOrUpdate(client *client.Client, hostName string, labels map[string]string) (*api.Node, error) {
+	n, err := Create(client, hostName, labels)
+	if err == nil {
+		return n, nil
+	}
+	if !errors.IsAlreadyExists(err) {
+		return nil, fmt.Errorf("unable to register %s with the apiserver: %v", hostName, err)
+	}
+
+	// fall back to update an old node with new labels
+	n, err = client.Nodes().Get(hostName)
+	if err != nil {
+		return nil, fmt.Errorf("error getting node %q: %v", hostName, err)
+	}
+	if n == nil {
+		return nil, fmt.Errorf("no node instance returned for %q", hostName)
+	}
+	return Update(client, n, labels)
+}
+
+// IsSlaveAttributeLabel returns true iff the given label is derived from a slave attribute
+func IsSlaveAttributeLabel(l string) bool {
+	return strings.HasPrefix(l, labelPrefix)
+}
+
+// IsUpToDate returns true iff the node's slave labels match the given attributes
+func IsUpToDate(n *api.Node, attrs []*mesos.Attribute) bool {
+	slaveLabels := map[string]string{}
+	labels := SlaveAttributesToLabels(attrs)
+	for k, v := range n.Labels {
+		if IsSlaveAttributeLabel(k) {
+			slaveLabels[k] = v
+		}
+	}
+	return reflect.DeepEqual(slaveLabels, labels)
+}
+
+// SlaveAttributesToLabels converts slave attributes into string key/value labels
+func SlaveAttributesToLabels(attrs []*mesos.Attribute) map[string]string {
+	l := map[string]string{}
+	for _, a := range attrs {
+		if a == nil {
+			continue
+		}
+
+		var v string
+		k := labelPrefix + a.GetName()
+
+		switch a.GetType() {
+		case mesos.Value_TEXT:
+			v = a.GetText().GetValue()
+		case mesos.Value_SCALAR:
+			v = strconv.FormatFloat(a.GetScalar().GetValue(), 'G', -1, 64)
+		}
+
+		if !validation.IsQualifiedName(k) {
+			log.V(3).Infof("ignoring invalid node label name %s: %s", k, v)
```

Comment:
nit: can we use `%q` instead of `%s` when printing k, v? may help w/ debugging..